pkg/updatestrategy/aws_ec2.go

```diff
@@ -0,0 +1,183 @@
+// the given clusterID and AWS session and.
+func NewEC2NodePoolBackend(clusterID string, sess *session.Session) *EC2NodePoolBackend {
+	return &EC2NodePoolBackend{
+		ec2Client: ec2.New(sess),
+		clusterID: clusterID,
+	}
+}
+
+// Get gets the EC2 instances matching to the node pool by looking at node pool
+// tag.
+// The node generation is set to 'current' for nodes with up-to-date
+// userData,ImageID and tags and 'outdated' for nodes with an outdated
+// configuration.
+func (n *EC2NodePoolBackend) Get(nodePool *api.NodePool) (*NodePool, error) {
+	instances, err := n.getInstances(nodePool)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list EC2 instances of the node pool: %w", err)
+	}
+
+	nodes := make([]*Node, 0)
+	for _, instance := range instances {
+		instanceID := aws.StringValue(instance.InstanceId)
+		node := &Node{
+			ProviderID:    fmt.Sprintf("aws:///%s/%s", aws.StringValue(instance.Placement.AvailabilityZone), instanceID),
+			FailureDomain: aws.StringValue(instance.Placement.AvailabilityZone),
+			Generation:    currentNodeGeneration,
+			// not used in clc logic
+			// Ready: true,
+		}
+
+		nodes = append(nodes, node)
+	}
+
+	// We only set Generation and Nodes as nothing else is needed by the
+	// CLC strategy
+	return &NodePool{
+		Generation: currentNodeGeneration,
+		Nodes:      nodes,
+	}, nil
+}
+
+// getInstances lists all running instances of the node pool.
+func (n *EC2NodePoolBackend) getInstances(nodePool *api.NodePool) ([]*ec2.Instance, error) {
+	params := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("tag:" + clusterIDTagPrefix + n.clusterID),
+				Values: []*string{
+					aws.String(resourceLifecycleOwned),
+				},
+			},
+			{
+				Name: aws.String("tag:" + nodePoolTag),
+				Values: []*string{
+					aws.String(nodePool.Name),
+				},
+			},
+		},
+	}
+
+	instances := make([]*ec2.Instance, 0)
+	err := n.ec2Client.DescribeInstancesPagesWithContext(context.TODO(), params, func(output *ec2.DescribeInstancesOutput, lastPage bool) bool {
+		for _, reservation := range output.Reservations {
+			for _, instance := range reservation.Instances {
+				switch aws.StringValue(instance.State.Name) {
+				case ec2.InstanceStateNameRunning, ec2.InstanceStateNamePending, ec2.InstanceStateNameStopped:
+					instances = append(instances, instance)
+				}
+			}
+		}
+		return true
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return instances, nil
+}
+
+type InstanceConfig struct {
+	UserData string
+	ImageID  string
+	Tags     map[string]string
+}
+
+// instanceConfigEqual compares current and desired InstanceConfig. It compares
+// userdata, imageID and checks if the current config has all the desired tags.
+// It does NOT check if the current config has too many EC2 tags as many tags are
+// injected out of our control. This means removing a tag is not enough to
+// make the configs unequal.
+func instanceConfigEqual(current, desired *InstanceConfig) bool {
```

Comment:
This function is not called. I think it should be called as part of the `Get` method to filter new/old nodes?